### PR TITLE
Allow using a laser only connected to a PWM pin, without an enable pin

### PR DESCRIPTION
**Pull request by ruediste**
### Description

I am using a laser module connected to a single PWM pin without a separate enable pin. This PR allows this configuration by allowing the absence of the SPINDLE_LASER_ENA_PIN.

### CODE REVIEW

**Comment by kaixxx** [comment]
Nice, I hope this gets merged. There are so many diode lasers out there without an enable pin (like mine)!

**Comment by InsanityAutomation** [comment]
Explicitly bypassing the Enable pin is a safety concern and should not be allowed to happen unintentionally. IF it is accepted to allow it to be bypassed this easily, I highly advise only doing so with an explicit configuration option warning of the risk of accidental laser pulsing, especially during power up of many boards, if an enabling device is not used. Often this is an SSR to the diode power or even on the input line as a safety cutoff. Not often built into the laser itself.

**Comment by tiagofreire-pt** [comment]
This seems to be the usecase of the Creality CV Laser unit for the Ender 3 S1.

https://github.com/CrealityOfficial/Ender-3S1/blob/main/Marlin/src/pins/stm32f4/pins_CREALITY_S1_F401RC.h#L231

**Comment by kaixxx** [comment]
> the risk of accidental laser pulsing, especially during power up of many boards, if an enabling device is not used.

The enable pin can also be in an undefined state (i.e. floating) on startup. You MUST use a pulldown resistor on both pins anyway (enable + PWM), so I don't see a real gain in security by making an enable pin mandatory.

**Comment by tiagofreire-pt** [comment]
> > the risk of accidental laser pulsing, especially during power up of many boards, if an enabling device is not used.
> 
> The enable pin can also be in an undefined state (i.e. floating) on startup. You MUST use a pulldown resistor on both pins anyway (enable + PWM), so I don't see a real gain in security by making an enable pin mandatory.

I'm not sure, but the `FAN_KICKSTART_TIME` and `FAN_MIN_PWM` shouldn't be set by default to `0` at the `configuration_adv.h`?

Such as:

```cpp
// When first starting the main fan, run it at full speed for the
// given number of milliseconds.  This gets the fan spinning reliably
// before setting a PWM value. (Does not work with software PWM for fan on Sanguinololu)

#if ENABLED(LASER_SYNCHRONOUS_M106_M107) || ENABLED(LASER_FEATURE) // tiagofreire-pt CV-Support
  #define FAN_KICKSTART_TIME 0
  #define FAN_MIN_PWM 0
#else
  #define FAN_KICKSTART_TIME 200
#endif
```

**Comment by ruediste** [comment]
So, what is the conclusion here? Should I add another configuration option like `SPINDLE_LASER_ALLOW_NO_ENA_PIN`?

**Comment by EvilGremlin** [comment]
better warning message like this https://github.com/MarlinFirmware/Marlin/blob/fc287b44da27d435e588f7d83ae61dcb4e929a5b/Marlin/src/inc/Warnings.cpp#L726
not adding line in configs

**Comment by descipher** [comment]
There is a new safety feature that will shutdown the laser if no movement occurs within a watchdog time period. Simply defining the pin used or not would have the same effect as bypassing the ENA circuit signal since its output is not used.

**Comment by descipher** [comment]
#22690 is now committed so there are many option’s available that were not previously possible. You can switch modes without recompiling. Fan sync mode is still compile based and is the same as pre #22690 but really is only required for boards with minimal pins. Sync can be done with M3 now or use inline mode. See https://marlinfw.org/docs/configuration/2.0.9/laser_spindle.html

**Comment by cbagwell** [comment]
PR worked well for me.  Allowed me to move my 2-pin Creality laser away from M106 with no safety nets to enabling LASER_FEATURE with watchdog timer that shuts off the laser with no movement.  Overall both an improvement in safety and features for my use case.

**Comment by cbagwell** [comment]
I'm continuing to test LASER_FEATURE with my 2 pin laser and think I've found an additional update needed to help in this case with no enable pin.

I'm adding some custom code in addition to this PR to attempt to speed up testing with M106 and M4 without reloading firmware to track down some differences in behavior I'm seeing.  My custom code had an issue I traced to SpindleLaser::set_enable() for CUTTER_MODE_STANDARD case.  I think the same issue might explain some M4 things I'm seeing even with just this PR.

Its invoking apply_power() before setting enable_state so its behavior is based on the old flag.  In disable case and this PR, this means it won't turn off laser.  This is not a big deal for most cases since other code tends to call apply_power() an addition time afterwards.  I think there are a few corner cases like menu options or safety shutdowns where this extra apply_power() might not occur though so would be best to have apply_power() honor enable flag.

Similar issue in same function for CUTTER_MODE_ERROR is it intends to force to enable_state = false but that gets reverted at end of function.  Probably not an issue because only time I see CUTTER_MODE_ERROR is followed by a set_enable(false) call.

Both issues can be addressed by moving enable_state = enable to top of function.

```
static void set_enabled(const bool enable) {
    switch (cutter_mode) {
      case CUTTER_MODE_STANDARD:
        apply_power(enable ? TERN(SPINDLE_LASER_USE_PWM, (power ?: (unitPower ? upower_to_ocr(cpwr_to_upwr(SPEED_POWER_STARTUP)) : 0)), 255) : 0);
        break;
      case CUTTER_MODE_CONTINUOUS:
        TERN_(LASER_FEATURE, set_inline_enabled(enable));
        break;
      case CUTTER_MODE_DYNAMIC:
        TERN_(LASER_FEATURE, set_inline_enabled(enable));
        break;
      case CUTTER_MODE_ERROR: // Error mode, no enable and kill power.
        enable_state = false;
        apply_power(0);
    }
    #if SPINDLE_LASER_ENA_PIN
      WRITE(SPINDLE_LASER_ENA_PIN, enable ? SPINDLE_LASER_ACTIVE_STATE : !SPINDLE_LASER_ACTIVE_STATE);
    #endif
    enable_state = enable;
}
```

File: Marlin/src/feature/spindle_laser.cpp [hunk 1]
```diff
@@ -67,7 +67,7 @@ cutter_frequency_t SpindleLaser::frequency;                           // PWM fre
 void SpindleLaser::init() {
   #if ENABLED(SPINDLE_SERVO)
     servo[SPINDLE_SERVO_NR].move(SPINDLE_SERVO_MIN);
-  #else
```
Comment (descipher) [inline Marlin/src/feature/spindle_laser.cpp:70]: This looks wrong to me, the #else is change to if and the remainder of the conditions are skipped when an ENA pin is not defined.

Comment (descipher) [reply Marlin/src/feature/spindle_laser.cpp:70]: Never mind, did not think it through.

Comment (descipher) [inline Marlin/src/feature/spindle_laser.cpp:70]: I would recommend something like this in conditionals

conditionals_adv.h
```
  #if PIN_EXISTS(SPINDLE_LASER_ENA)
    #define HAS_SPINDLE_LASER_ENA 1
  #endif
```
Then use TERN_ elsewhere e.g.

```TERN_(HAS_SPINDLE_LASER_ENA, OUT_WRITE(SPINDLE_LASER_ENA_PIN, !SPINDLE_LASER_ACTIVE_STATE));    // Init spindle to off```

**Comment by descipher** [comment]
Actually it does not get reverted at the end, it is set to the parameter enable which is either true or false
e.g. call set_enabled(false)
```
 static void set_enabled(const bool enable) {

     enable_state = enable // enable is false
```
Error condition will be reverted .. not really it sets power to 0 but it should return at that point to avoid enabling, moving enable_state to the top looks fine to me.

**Comment by cbagwell** [comment]
> Actually it does not get reverted at the end, it is set to the parameter enable which is either true or false e.g. call set_enabled(false)
> 
> ```
>  static void set_enabled(const bool enable) {
> 
>      enable_state = enable // enable is false
> ```
> 
> Error condition will be reverted .. not really it ses power to 0 but it should return at that point to avoid enabling, moving enable_state to the top looks fine to me.

Thanks for reviewing and commenting.  I also happen to log some SERIAL_ECHO's that help show what I'm talking about and to see if alternative changes are better.  I added echos to set_enable(), apply_power(), and _set_ocr() to see if laser was enabled/disabled when I thought it should be and I used laser toggle menu to enable/disable laser.

First couple of apply_power() are examples from periodic apply_power() from planner.  Next is I set power to 10 and enabled by menu toggle but notice enabled() is still false so _set_ocr(0) occurs.   We might consider that a feature since that 204 value appears to be garbage data.  

In code paths of menu toggle, G1, and M3-M5 there is a forced call to apply_power() right after any set_enable(true) and so that is what does real power setting in today's logic... and the periodic apply_apply() catches most other corner cases if they missed an apply_power.

For toggle off case, its shutting off because power==0 since enabled() is true still.  I think there is less cases that do a force apply_power() after set_enabled(false) so maybe more important to shutdown immediately.

```
apply_power opwr: 0 enabled(): 0
apply_power opwr: 0 enabled(): 0  <--- periodic
set_enabled enable: 1
apply_power opwr: 204 enabled(): 0 <-- called by set_enabled from menu toggle
_set_ocr ocr: 0  <--- better value than 204
apply_power opwr: 10 enabled(): 1 <--- called from menu toggle after set_enabled call
_set_ocr ocr: 10
apply_power opwr: 10 enabled(): 1 <--- periodic again
apply_power opwr: 10 enabled(): 1
[continues till I toggle laser off. ]
set_enabled enable: 0
apply_power opwr: 0 enabled(): 1 <--- called by set_enabled from menu toggle
_set_ocr ocr: 0 <--- still comes to right answer.
```

So I can think of two options.  1) I can track down source of garbage and make sure its a meaningful value always before calling set_enabled() and then move variable setting so apply_power() is able to turn on laser or 2) Save some time by only calling apply_power() in STANDARD mode for enable == false case and force everyone else to make a call to apply_power afterwards for set_enable(true) cases.

**Comment by descipher** [comment]
Thank you for testing it!

The 204 value is correct and appropriate, it is 80% laser power when using PWM255 as a power unit. If you use the LCD menu and turn on the laser when 0 power is set it will apply at 80% by default. This is a previous behaviour prior to the recent laser PR's. 
```
  /**
   * Speed / Power can be set ('M3 S') and displayed in terms of:
   *  - PWM255  (S0 - S255)
   *  - PERCENT (S0 - S100)
   *  - RPM     (S0 - S50000)  Best for use with a spindle
   *  - SERVO   (S0 - S180)
   */
  #define CUTTER_POWER_UNIT PWM255
```
Keep in mind that you cannot just disable the laser at any point as the different modes have their own states to deal with.
The issue I see at the moment is the error state. We need to move enabled_state to the beginning of set_enabled to prevent the error set enable_state = false being overridden by any set_enabled = true. This is the simplest change to address that issue.

There are many past historical modes of op that need to be kept intact if possible. e.g. STANDARD_MODE is the SPINDLE compatible op. when in this mode SPINDLE ops are functional as basic M3/M4/M5 so when someone uses a SPINDLE it will continue to work as previously defined. This also applies to the inline mode of op and Software products such as LaserGRBL LightBurn LaserWEB and so forth.  

To work in single pin mode we may need to consider additional enable_state conditions to keep the previous compatibility intact.    

Oh .. note: The periodic behaviour will be the menu refresh rate ...

**Comment by cbagwell** [comment]
OK, then to summarize the subthread.

* I thought a code change might be needed to prevent unwanted power values from being used that were previously hidden by PIN enable/disable.
* After discussion and further testing, turns out the behavior will be same for both cases of pin and no pin.
* There might be room for improvements to set_enable and can be made in this PR or some other one since they are not related to using lasers without enable pin.
* Using M106 with Fan Sync is better option than using M4 in STANDARD mode since later is meant more for spindles that are normally always left on and often at a single speed always... so I should move my laser testing over to inline modes now.

**Comment by descipher** [comment]
Not sure why to use M4 at all, generally inline will be the most useful method for a laser. M106 allows a user to leave everything in place and just plug into a PWM based FAN as the output control.

see https://marlinfw.org/docs/configuration/2.0.9/laser_spindle.html

**Comment by descipher** [comment]
> This seems to be the usecase of the Creality CV Laser unit for the Ender 3 S1.
> 
> https://github.com/CrealityOfficial/Ender-3S1/blob/main/Marlin/src/pins/stm32f4/pins_CREALITY_S1_F401RC.h#L231

That uses Marlin 2.0.8.1 .. just a wee tad out of date .. lol

**Comment by cbagwell** [comment]
> Not sure why to use M4 at all, generally inline will be the most useful method for a laser. M106 allows a user to leave everything in place and just plug into a PWM based FAN as the output control.
> 
> see https://marlinfw.org/docs/configuration/2.0.9/laser_spindle.html

I really meant to say M3 instead of M4 but mostly same difference.  

I am working from left to right in Lightburn's device setting screen.  I have my laser hooked to FAN0 and so started with M106/M107 setting.  Then I recompiled with LASER_FEATURE so I could see how Lightburn M3/M5 setting worked.  I had some issues with M3/M5 so moved to Lightburn Inline and that was worse in that laser never turned on at all.  So I circled back to M3/M5 for a while.

I've debugged enough now I know why Inline didn't work.  Its because you need to send an M3 I before using Lightburn's inline support or else G1 Sxx won't turn on laser while in Standard mode.  Marlin docs are pretty clear about needing to do that but not easy for a new person to connect dots why Lightburn's Inline option wasn't taking care of that for me.

I don't know Marlin history to know why "G1 I" is a thing but if perhaps it could not only call set_enable() to be backwards compatible with LightBurn but also switch to Continuous Inline Mode if in Standard Mode then life would be sweet with LightBurn and its Marlin Inline setting.  Or perhaps LightBurn developer can be convinced to update to send M3 I or M4 I to align with recent Marlin behavior.

**Comment by descipher** [comment]
Noted: I will update him on the new capability.

File: Marlin/src/feature/spindle_laser.cpp (hunk 1 above)
Comment (thinkyhead) [reply Marlin/src/feature/spindle_laser.cpp:70]: Unfortunately, I don't think all the versions of `WRITE` are happy when used inside of `TERN` macros. But other than FastIO macros most things work ok there.

**Pull request merged**